zeus/device/cpu/rapl.py

```diff
@@ -0,0 +1,180 @@
+from zeus.utils.logging import get_logger
+
+logger = get_logger(name=__name__)
+
+RAPL_DIR = "/sys/class/powercap/intel-rapl"
+
+
+def rapl_is_available() -> bool:
+    """Check if RAPL is available."""
+    if not os.path.exists(RAPL_DIR):
+        logger.info("RAPL is not supported on this CPU.")
+        return False
+    logger.info("RAPL is available.")
+    return True
+
+
+class ZeusRAPLNotSupportedError(ZeusBaseCPUError):
+    """Zeus CPU exception class Wrapper for RAPL not supported on CPU."""
+
+    def __init__(self, message: str) -> None:
+        """Initialize Zeus Exception."""
+        super().__init__(message)
+
+
+class ZeusRAPLFileInitError(ZeusBaseCPUError):
+    """Zeus CPU exception class Wrapper for RAPL file initialization error on CPU."""
+
+    def __init__(self, message: str) -> None:
+        """Initialize Zeus Exception."""
+        super().__init__(message)
+
+
+class RAPLFile:
+    """RAPL File class for each RAPL file.
+
+    This class defines the interface for interacting with a RAPL file for a package. A package can
+    be a CPU or DRAM
+    """
+
+    def __init__(self, path: str) -> None:
+        """Initialize RAPL file object. Each RAPL file object manages one energy_uj file."""
+        self.path = path
+        self.energy_uj_path = os.path.join(path, "energy_uj")
+        try:
+            with open(os.path.join(path, "name"), "r") as name_file:
+                self.name: str = name_file.read().strip()
+        except FileNotFoundError as err:
+            raise ZeusRAPLFileInitError("Error reading package name") from err
+        try:
+            with open(self.energy_uj_path) as energy_file:
+                self.last_energy = int(float(energy_file.read().strip()) / 10**3)
+        except FileNotFoundError as err:
+            raise ZeusRAPLFileInitError("Error reading package energy") from err
+        try:
+            with open(
+                os.path.join(path, "max_energy_range_uj"), "r"
+            ) as max_energy_file:
+                self.max_energy_range_uj = int(
+                    float(max_energy_file.read().strip()) / 10**3
+                )
+        except FileNotFoundError as err:
+            raise ZeusRAPLFileInitError(
+                "Error reading package max energy range"
+            ) from err
+
+    def __str__(self) -> str:
+        """Return a string representation of the RAPL file object."""
+        return f"Path: {self.path}\nEnergy_uj_path: {self.energy_uj_path}\nName: {self.name}\
+        \nLast_energy: {self.last_energy}\nMax_energy: {self.max_energy_range_uj}"
+
+    def _read(self) -> int:
+        """Read the current energy value from the energy_uj file.
+
+        Returns:
+            int: The current energy value in millijoules.
+        """
+        with open(self.energy_uj_path) as energy_file:
+            self.last_energy = int(float(energy_file.read().strip()) / 10**3)
+        return self.last_energy
+
+    def _read_delta(self) -> int:
```

Comment:
1. Is it possible for the `energy_uj` counter to max out and reset *more than once* between two readings? How should that be detected?
2. Where is this function used? Dealing with wrap-arounds seems important but I can't seem to find where it's being called.
3. You can drop the type hints `: int` for both `last_energy` and `new_energy`.

Reply:
1. Right now I can only think of a way to detect overflows if it only happens once and if wraps around to a value lower than `self.last_energy`. The only way I can think of keeping track of wrap around would be to read from the file periodically and keep track of how many times a new reading becomes less than `self.last_energy`.
2. `_read_delta` isn't used anywhere, it was initially for testing the `RAPLFile` class that I forgot to remove. I'll moved the wrap around logic to `read`

Reply:
For 1, please file an issue. The solution makes sense.